pal_event_server/lib/src/allowlist.dart

```diff
@@ -14,129 +14,193 @@
+        return _appForContentRuleExpression.hasMatch(appsUsedValue) &&
+            _expression.hasMatch(appContentValue);
+      default:
+        false;
     }
 
-    return _hideAllButAppName(appName);
   }
 
-  List<Map<String, dynamic>> filterData(List eventJson) {
-    final results = <Map<String, dynamic>>[];
-    for (var event in eventJson) {
-      // TODO This probably shouldn't be here?
-      if (event['experimentGroupName'] != 'AppLog') {
-        results.add(event);
-      } else {
-        var appName = '';
-        var appsUsedRaw = '';
-        var appContent = '';
-        var appUrl = '';
-
-        final responsesAllowlisted = <Map<String, dynamic>>[];
-        final responses = event['responses'];
-        for (Map<String, dynamic> response in responses) {
-          final responseName = response['name'];
-          final responseAnswer = response['answer'];
-
-          if (responseName == 'apps_used') {
-            appName = responseAnswer;
-          } else if (responseName == 'apps_used_raw') {
-            appsUsedRaw = responseAnswer;
-          } else if (responseName == 'app_content') {
-            appContent = responseAnswer;
-          } else if (responseName == 'url') {
-            appUrl = responseAnswer;
-          } else {
-            responsesAllowlisted.add(response);
-          }
-        }
+    @override
+    String toString() {
+      return 'AllowListRule{_type: $_type, _expression: $_expression, _appForContentRule: $_appForContentRule}';
+    }
+
+    static AllowListRule ofAppUsed(String expression) {
+      return AllowListRule({"type": APPS_USED_RULE_TYPE, "expression": expression});
+    }
+
+    static AllowListRule ofAppContent(String app, String expression) {
+      return AllowListRule({"type": APP_CONTENT_RULE_TYPE, "app": app, "expression": expression});
+    }
+  }
+
+
+class AllowList {
+  var _rules = <AllowListRule>[];
+
+  Iterable<AllowListRule> _appRules;
+
+  Iterable<AllowListRule> _appContentRules;
+
+  get rules => _rules;
+
+  set rules(value) {
+    _rules = value;
+    _appRules = _rules.where((element) => element._type == APPS_USED_RULE_TYPE);
+    _appContentRules =
+        _rules.where((element) => element._type == APP_CONTENT_RULE_TYPE);
+  }
+
+  List<Event> filterData(List<Event> events) {
+    for (var event in events) {
+      wipeDetailsOnEvent(event);
+      filter(event);
+    }
+    return events;
+  }
 
-        final data = _filterData(appName, appsUsedRaw, appContent, appUrl);
+  filter(Event event) {
+    _logger.info("Event: ${event.toJson()}");
+    if (event.groupName != 'APPUSAGE_DESKTOP') {
+      return event;
+    }
+    if (_appRules == null) {
+      hashAllAppLoggerFields(event);
+      return;
+    }
+    var allowed = false;
+    for (var appRule in _appRules) {
+      if (event.responses.containsKey(appsUsedKey) &&
+          event.responses[appsUsedKey] != null &&
+          appRule.matches(event.responses[appsUsedKey], null)) {
```

Comment:
nit: We can use syntax sugar here
```
if (appRule.matches(event.responses[appsUsedKey] ?? '', null))
```
since for a map `operator[]` will return null if key is not found.

Reply:
Nice. Done.